overlord/snapstate/handlers.go

```diff
@@ -3538,14 +3665,58 @@ func (m *SnapManager) doSetupAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 	curAliases := snapst.Aliases
+	autoDisabled := snapst.AutoAliasesDisabled
 
-	_, _, err = applyAliasesChange(snapName, autoDis, nil, snapst.AutoAliasesDisabled, curAliases, m.backend, doApply)
+	prune, oldAliases, oldAutoDisabled, err := shouldPruneOldAliases(t, snapst)
 	if err != nil {
 		return err
 	}
 
+	t.Set("old-aliases-pruned", prune)
+	// AliasesPending must be set to false before applying changes so that the
+	// undo for set-auto-aliases can revert aliases on disk if this task fails
+	// due to applyAliasesChange failing mid-way.
 	snapst.AliasesPending = false
 	Set(st, snapName, snapst)
+
+	// no need to check for conflicts as it was already checked in `set-auto-aliases`
+	_, _, err = applyAliasesChange(snapName, oldAutoDisabled, oldAliases, autoDisabled, curAliases, m.backend, doApply)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *SnapManager) undoSetupAliases(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.InstanceName()
+
+	var oldAliasesPruned bool
+	if err := t.Get("old-aliases-pruned", &oldAliasesPruned); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if oldAliasesPruned {
+		// AliasesPending is set to false so that the undo for set-auto-aliases
+		// can revert aliases on disk.
+		snapst.AliasesPending = false
```

Comment:
shouldn't this already be the case anyway?

Reply:
yes, removed